### PR TITLE
2.7 Upgrade for Systemd File Relocation

### PR DESCRIPTION
**Pull request by manadart**
### Checklist

 - [x] Checked if it requires a [pylibjuju](https://github.com/juju/python-libjuju) change?
 - [x] Added [integration tests](https://github.com/juju/juju/tree/develop/tests) for the PR?
 - [x] Added or updated [doc.go](https://discourse.jujucharms.com/t/readme-in-packages/451) related to packages changed?
 - [x] Do comments answer the question of why design decisions were made?

----

## Description of change

This patch adds a 2.7.2 upgrade step to relocate systemd unit files for agents.

The `UpgradableService` indirection has a new method `RemoveOldService`, which is implemented by `systemd.Service`. It removes the legacy service directory from _/lib/systemd/system_.

This method is called by the service manager when writing files, so it will work for all prior upgrade steps using the service manager to rewrite service units.

Refactoring includes adding a shim for file-system operations and removing some of the patching-mocks-over-function-vars messiness.

## QA steps

- Bootstrap with a prior Juju (2.6 will do).
- `juju deploy cs:~jameinel/ubuntu-lite`
- When settled, build this patch and run `juju upgrade-controller --build-agent`.
- `juju upgrade-model`.
- Check that everything is online.
- `juju ssh 0`. Check that _/etc/systemd/system_ contains service definitions and start scripts for the unit and machine, and there are no directories for them in _/lib/systemd/system_.

I also successfully tested a series upgrade from Trusty to Xenial, but this should not be required - it is covered by existing functional tests.

## Documentation changes

None.

## Bug reference

https://bugs.launchpad.net/juju/+bug/1860432

### CODE REVIEW

File: service/common/testing/fake.go [hunk 1]
```diff
@@ -268,24 +264,25 @@ func (ss *FakeService) Remove() error {
...
-	retErr := ss.NextErr()
-	if retErr != nil {
-		return retErr
-	}
-	return nil
+	return ss.simpleCall("WriteService")
+}
+
+// RemoveOldService implements UpgradableService.
+func (ss *FakeService) RemoveOldService() error {
```
Comment (manadart) [inline service/common/testing/fake.go]: I would like to get rid of these fakes, but I already undertook some mechanical improvement work in this patch.

File: service/systemd/service.go [hunk 2]
```diff
@@ -69,27 +68,23 @@ type Service struct {
...
 // NewService returns a new reference to an object that implements the Service
 // interface for systemd.
 func NewService(
-	name string, conf common.Conf, dataDir string, newDBus DBusAPIFactory, fallBackDirName string,
+	name string, conf common.Conf, dataDir string, newDBus DBusAPIFactory, fileOps FileOps, fallBackDirName string,
 ) (*Service, error) {
 	confName := name + ".service"
-	var volName string
-	if conf.ExecStart != "" {
-		volName = renderer.VolumeName(common.Unquote(strings.Fields(conf.ExecStart)[0]))
```
Comment (manadart) [inline service/systemd/service.go]: `VolumeName` always returns "" for *nix, and we know that systemd is always *nix, so this is not required.

**Comment by manadart** [comment]
[Upgrade Paths for Systemd Relocation](https://trello.com/c/NGB613ef/2257-upgrade-paths-for-systemd-relocation)

File: service/agentconf_test.go [hunk 3]
```diff
@@ -316,9 +316,13 @@ func (s *agentConfSuite) TestWriteSystemdAgentsSystemdNotRunning(c *gc.C) {
...
 
 func (s *agentConfSuite) TestWriteSystemdAgentsDBusErrManualLink(c *gc.C) {
+	// nil errors are for calls to RemoveOldService.
 	s.services[0].SetErrors(
+		nil,
 		errors.New("No such method 'LinkUnitFiles'"),
+		nil,
 		errors.New("No such method 'LinkUnitFiles'"),
+		nil,
 		errors.New("No such method 'LinkUnitFiles'"),
```
Comment (SimonRichardson) [inline service/agentconf_test.go]: This is quite horrid.

```go
err := errors.New("no such method 'LinkUnitFiles'")
s.services[0].SetErrors(nil, err, nil, err, nil, err)
```

File: service/systemd/export_test.go [hunk 4]
```diff
@@ -30,13 +30,6 @@ func PatchNewChan(patcher patcher) chan string {
 	return ch
 }
 
-func PatchFileOps(patcher patcher, ctrl *gomock.Controller) *MockShimFileOps {
-	mock := NewMockShimFileOps(ctrl)
-	patcher.PatchValue(&remove, mock.Remove)
-	patcher.PatchValue(&createFile, mock.CreateFile)
-	return mock
-}
```
Comment (SimonRichardson) [inline service/systemd/export_test.go]: Lovely!

File: service/systemd/interface.go [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+	EnableUnitFiles([]string, bool, bool) (bool, []dbus.EnableUnitFileChange, error)
+	DisableUnitFiles([]string, bool) ([]dbus.DisableUnitFileChange, error)
+	GetUnitProperties(string) (map[string]interface{}, error)
+	GetUnitTypeProperties(string, string) (map[string]interface{}, error)
+	Reload() error
+}
+
+// FileOps describes file-system operations required to install
+// and remove a service.
+type FileOps interface {
```
Comment (SimonRichardson) [inline service/systemd/interface.go]: Nit: FileSystemOps.

Comment (manadart) [reply service/systemd/interface.go]: Changed it. Implementation too.

File: service/agentconf_test.go (hunk 3 above)
Comment (manadart) [reply service/agentconf_test.go]: Yep, done.

**Comment by manadart** [comment]
$$merge$$

**Pull request merged**